### PR TITLE
refactor(buildPrettierOptions.js): Exclusively use local config or pl…

### LINKED ISSUE #370: Request: remove merging of explicit configurations with atom-prettier settings
I just spent some time looking into why my prettier settings in package.json weren't being respected, and it's due to [them being assigned over atom-prettier settings](https://github.com/prettier/prettier-atom/blob/master/src/executePrettier/buildPrettierOptions.js#L70-L74).  I'm wondering what the use-case is for

1. allowing users to `only format if a prettier config is found` and
2. assigning the prettier config over those in atom-prettier.

i.e. why would I want to require an explicit configuration while also using the one local to my editor?  I assume the use-case for allowing users to require both an explicit prettier instance and settings would be to ensure consistent formatting should they move to a different editor or machine.

Even if I set atom-prettier's settings to the defaults, this could become problematic should the defaults in the local prettier instance be different.

Another less breaking solution would be to add an option to ignore atom-prettier's configuration settings.

I'm willing to create a PR if either solution is agreeable.

### PR DESCRIPTION
**Pull request by olsonpm**
…ugin settings

Stop the current behavior of merging a local prettier config with
  those found in the plugin settings.  The reason being a project
  should be formatted according to its own rules, and per Prettier
  convention, omitted rules should fallback to defaults.

Remove the plugin's integration with editorconfig in favor of relying
  on prettier's via its editorconfig option.

BREAKING CHANGE: The formatting configuration is now built according to different rules

fix #370

### CODE REVIEW

File: src/executePrettier/buildPrettierOptions.test.js [hunk 1]
```diff
@@ -109,40 +107,46 @@ it('uses vue as the parser if current scope is listed as a Vue SFC scope in sett
...
-  shouldUseEditorConfig.mockImplementation(() => false);
-
   buildPrettierOptions(editor);
+  expect(mockPrettierInstance.resolveConfig.sync.mock.calls[0][1]).toEqual({ editorconfig });
 
-  expect(buildEditorConfigOptions).not.toHaveBeenCalled();
+  editorconfig = true;
+  buildPrettierOptions(editor);
+  expect(mockPrettierInstance.resolveConfig.sync.mock.calls[1][1]).toEqual({ editorconfig });
 });
```
Comment (robwise) [inline src/executePrettier/buildPrettierOptions.test.js]: Can we either:

1. split this into two tests (one for `true` and one for `false`)
2. just test one of the two cases since it shouldn't make a difference whether it's `true` or `false`

(I'd prefer the latter). This way we aren't mutating a variable and doing multiple expectations in a single test.

File: src/executePrettier/buildPrettierOptions.test.js [hunk 2]
```diff
@@ -109,40 +107,46 @@ it('uses vue as the parser if current scope is listed as a Vue SFC scope in sett
...
 
-it('overrides values with editorconfig values if editor config is enabled', () => {
+it('only uses project prettier config when found', () => {
   const editor = buildMockEditor();
   const fakePrettierOptions = { tabWidth: 2, printWidth: 80, useTabs: false };
```
Comment (robwise) [inline src/executePrettier/buildPrettierOptions.test.js]: let's strike this line since it shouldn't be used

File: src/executePrettier/buildPrettierOptions.test.js [hunk 3]
```diff
@@ -109,40 +107,46 @@ it('uses vue as the parser if current scope is listed as a Vue SFC scope in sett
...
-  const fakeEditorConfigOptions = { tabWidth: 4, printWidth: 100, useTabs: true };
-  getPrettierOptions.mockImplementation(() => fakePrettierOptions);
-  buildEditorConfigOptions.mockImplementation(() => fakeEditorConfigOptions);
-  shouldUseEditorConfig.mockImplementation(() => true);
+  const fakePrettierConfigOptions = { tabWidth: 4, printWidth: 100, useTabs: true };
+  const mockPrettierInstance = { resolveConfig: { sync: jest.fn(() => fakePrettierConfigOptions) } };
+  getPrettierInstance.mockImplementation(() => mockPrettierInstance);
   getCurrentFilePath.mockImplementation(() => 'parent-dir/foo.js');
+  getPrettierOptions.mockImplementation(() => fakePrettierOptions);
```
Comment (robwise) [inline src/executePrettier/buildPrettierOptions.test.js]: let's strike this line since it shouldn't be used

File: src/executePrettier/buildPrettierOptions.test.js [hunk 4]
```diff
@@ -109,40 +107,46 @@ it('uses vue as the parser if current scope is listed as a Vue SFC scope in sett
...
+  expect(mockPrettierInstance.resolveConfig.sync.mock.calls[0][1]).toEqual({ editorconfig });
 
-  expect(buildEditorConfigOptions).not.toHaveBeenCalled();
+  editorconfig = true;
+  buildPrettierOptions(editor);
+  expect(mockPrettierInstance.resolveConfig.sync.mock.calls[1][1]).toEqual({ editorconfig });
 });
 
-it('overrides values with editorconfig values if editor config is enabled', () => {
+it('only uses project prettier config when found', () => {
```
Comment (robwise) [inline src/executePrettier/buildPrettierOptions.test.js]: This makes sense in the context of this PR, but in the future, some may not understand what we mean by" only uses project prettier config." I think we should change to `uses settings from the prettier config if found`, and then we can make the test below use `uses settings from prettier-atom if the prettier config is not found`.

File: src/executePrettier/buildPrettierOptions.test.js (hunk 1 above)
Comment (olsonpm) [reply src/executePrettier/buildPrettierOptions.test.js]: Makes sense.  I appreciate you explaining "why" - I honestly am not a heavy tester so I don't have much knowledge of conventions or best practices.

File: src/executePrettier/buildPrettierOptions.test.js (hunk 4 above)
Comment (olsonpm) [reply src/executePrettier/buildPrettierOptions.test.js]: good point

File: src/executePrettier/buildPrettierOptions.test.js (hunk 2 above)
Comment (olsonpm) [reply src/executePrettier/buildPrettierOptions.test.js]: whoops - thanks

**Comment by robwise** [comment]
> I can't think of a scenario where a user would want to use an .editorconfig file with prettier-atom settings.

Although I wouldn't recommend it because of large diff noise, perhaps if a user was working on a project that had not adopted prettier, but had adopted an editorconfig, and the user wants to use prettier for their own personal convenience? I would assume this is the use case that got the editorconfig integration put into the prettier repo in the first place. Because the only other use case is where a project adopts both prettier and an editorconfig but doesn't bother to make sure their prettier config and editor config agree.

So anyway, I feel like there's even more of a use case (if not more) for wanting to respect the editorconfig settings when using personal prettier-atom defaults than when using prettier by itself. Although, I completely overlooked this when first reviewing the PR. So maybe we _should_ keep the editorconfig stuff, but only apply it if a prettier config isn't found?

And the implementation logic would be that we take the editorconfig defaults as the priority (assuming the option to use editorconfig is checked), and then anything that isn't specified, we use the prettier-atom settings.

**Comment by olsonpm** [comment]
> the user wants to use prettier for their own personal convenience

Do you guys want to support that use-case?  Basically allowing someone to format a project file according to settings located on his local machine?  If so then that's totally fine - just not something I agree with.
<br>

> a project adopts both prettier and an editorconfig but doesn't bother to make sure their prettier config and editor config agree.

I see.  I agree it's odd to have a scenario where a prettier configuration overrides an editorconfig setting.  Seems to me this should be presented as an error for the user to fix.
<br>

> maybe we should keep the editorconfig stuff, but only apply it if a prettier config isn't found?

I'll do that.

**Comment by robwise** [comment]
> Do you guys want to support that use-case? Basically allowing someone to format a project file according to settings located on his local machine? If so that's totally fine - just not something I agree with.

This is basically the use case for adding editorconfig to the main prettier repo. Otherwise they could just rely on making sure the prettierrc was set correctly.

**Comment by olsonpm** [comment]
I guess I see it a little differently.  Prettier allows you to fall back to editorconfig defaults instead of prettier defaults.  e.g. this makes sense:

```
// .prettierrc
{
  singleQuote: true
}
```

```
// .editorconfig
[*.js]
indent_style = tab
indent_size = 4
```

It only gets confusing when a prettier config explicitly conflicts with an editorconfig.

**Comment by olsonpm** [comment]
Alright I believe it's working as you intend.  I realized I was unnecessarily building the atom settings object if it wasn't going to be used, so I refactored that into its own function.

I also incorporated a `describe` block to separate the 'config found' and 'config not found' conditions.  Please let me know if the tests should be organized differently.  Also I appreciate your patience.

**Comment by robwise** [comment]
> I guess I see it a little differently. Prettier allows you to fall back to editorconfig defaults instead of prettier defaults

It does allow you to do that sure, but that's not very useful IMO. The case I see much more frequently and I think the reason for them originally implementing this was as a convenience for editor integrations that wanted to support this type of thing. I believe prettier-atom was the only integration that had built its own editorconfig support at the time.

We've gotten lots of prettier-atom devs who either are not wanting to go through the trouble of having to create a config file (or even add prettier locally) or are not allowed to do so by their team (like Kent Dodds), yet still want to use prettier. If these projects do have an editorconfig, then their prettier needs to be compatible with those settings.

File: src/executePrettier/buildPrettierOptions.js [hunk 5]
```diff
@@ -70,8 +72,10 @@ const buildPrettierOptions = (editor: TextEditor) => {
   return {
     ...optionsFromSettings,
     ...buildEditorConfigOptionsIfAppropriate(editor),
-    ...getPrettierConfigOptions(editor),
   };
 };
 
+const buildPrettierOptions = (editor: TextEditor) =>
+  getPrettierConfigOptions(editor) || getOptionsFromSettings(editor);
```
Comment (robwise) [inline src/executePrettier/buildPrettierOptions.js]: Does this mean we ignore the user's parser configuration settings if there is a prettier config found? That might be confusing as those settings exist outside the "prettier defaults" section and don't have any indication that they will be ignored.

**Comment by olsonpm** [comment]
ah yeah you're right having the two configs doesn't seem useful.  Not being allowed to add prettier seems weird, but i'm sure it's for good reason.

File: src/executePrettier/buildPrettierOptions.test.js [hunk 6]
```diff
@@ -109,40 +109,60 @@ it('uses vue as the parser if current scope is listed as a Vue SFC scope in sett
...
+    getPrettierInstance.mockImplementation(() => mockPrettierInstance);
+    shouldUseEditorConfig.mockImplementation(() => false);
+    getCurrentFilePath.mockImplementation(() => 'parent-dir/foo.js');
+    getPrettierOptions.mockImplementation(() => fakePrettierOptions);
+
+    const actual = buildPrettierOptions(editor);
+
+    expect(buildEditorConfigOptions).not.toHaveBeenCalled();
+    expect(actual).toEqual(fakePrettierOptions);
+  });
```
Comment (robwise) [inline src/executePrettier/buildPrettierOptions.test.js]: - In the first test, let's add some line returns, so we have a block each for given, when, and then? This allows the reader to jump straight to the setup, or straight to the action under test, or straight to the expectation.
- Usually, when you test you just write the base case and then specify certain edge cases, that way you don't end up with all of your tests having to test that things **didn't** like crazy.
- This "base case" or "happy path" test should be the first test, and then other tests can just specify how they diverge from the path.

All together:

```js
describe('when a prettier config is found', () => {
  it('returns settings from the prettier config', () => {
    const editor = buildMockEditor();
    const fakePrettierConfigOptions = { tabWidth: 4, printWidth: 100, useTabs: true };
    const mockPrettierInstance = { resolveConfig: { sync: jest.fn(() => fakePrettierConfigOptions) } };
    getPrettierInstance.mockImplementation(() => mockPrettierInstance);
    getCurrentFilePath.mockImplementation(() => 'parent-dir/foo.js');

    const actual = buildPrettierOptions(editor);

    expect(actual).toEqual(fakePrettierConfigOptions);
  });
  
  it('tells prettier whether to use editorconfig when resolving options', () => {
    const editor = buildMockEditor();
    const mockPrettierInstance = { resolveConfig: { sync: jest.fn(() => ({})) } };
    getPrettierInstance.mockImplementation(() => mockPrettierInstance);
    shouldUseEditorConfig.mockImplementation(() => true);
    getCurrentFilePath.mockImplementation(() => 'parent-dir/foo.js');
    
    buildPrettierOptions(editor);
    
    expect(mockPrettierInstance.resolveConfig.sync.mock.calls[0][1]).toEqual({ editorconfig: true });
  });
});

describe('when a prettier config is not found', () => {
  it('returns settings from prettier-atom', () => {
    const editor = buildMockEditor();
    const fakePrettierOptions = { singleQuote: true, tabWidth: 2 };
    const mockPrettierInstance = { resolveConfig: { sync: jest.fn(() => null) } };
    getPrettierInstance.mockImplementation(() => mockPrettierInstance);
    shouldUseEditorConfig.mockImplementation(() => false);
    getCurrentFilePath.mockImplementation(() => 'parent-dir/foo.js');
    getPrettierOptions.mockImplementation(() => fakePrettierOptions);

    const actual = buildPrettierOptions(editor);

    expect(actual).toEqual(fakePrettierOptions);
  });

  it('overrides prettier-atom settings with editorconfig values if editor config is enabled', () => {
    const editor = buildMockEditor();
    const fakePrettierOptions = { singleQuote: true, tabWidth: 2 };
    const fakeEditorConfigOptions = { tabWidth: 4 };
    const mockPrettierInstance = { resolveConfig: { sync: jest.fn(() => null) } };
    getPrettierInstance.mockImplementation(() => mockPrettierInstance);
    buildEditorConfigOptions.mockImplementation(() => fakeEditorConfigOptions);
    shouldUseEditorConfig.mockImplementation(() => true);
    getCurrentFilePath.mockImplementation(() => 'parent-dir/foo.js');
    getPrettierOptions.mockImplementation(() => fakePrettierOptions);

    const actual = buildPrettierOptions(editor);

    expect(actual).toEqual({ ...fakePrettierOptions, ...fakeEditorConfigOptions });
  });
});
```

**Comment by SavePointSam** [comment]
A bit of background (as I've been digging through Prettier myself). `.editorconfig` options are overridden by Prettier settings. And `.editorconfig` is only picked up if the `editorconfig` flag is provided when doing configuration resolution. Meaning that you get the following hierarchy:

```
.prettierrc / package.json > .editorconfig > prettier defaults
```

I would suggest that the following pattern be observed for us:

```
.prettierrc / package.json > .editorconfig > prettier-atom settings > prettier defaults
```

It seems [my thoughts on the logic](https://github.com/prettier/prettier-atom/issues/370#issuecomment-362813322) here may have been lost in the shuffle. We are still only using the `prettier-atom` settings in specific use-cases right? (I haven't had time to review the PR, just this conversation.)

File: src/executePrettier/buildPrettierOptions.js (hunk 5 above)
Comment (SavePointSam) [reply src/executePrettier/buildPrettierOptions.js]: True, we only want to discard the prettier configuration settings explicitly, not every setting that `prettier-atom` has. In some advanced cases, a user may want to select a custom parser. We shouldn't take those things away, simply merge over them.

File: src/executePrettier/buildPrettierOptions.test.js (hunk 6 above)
Comment (SavePointSam) [reply src/executePrettier/buildPrettierOptions.test.js]: We may be able to utilize prettier for the `overrides prettier-atom settings with editorconfig values if editor config is enabled` case. If you ask for settings from prettier with `editorconfig` enabled, it should return an object where it found the editor config and combined those with prettier defaults for us. No need to implement a custom `editorconfig` solution.

**Comment by SavePointSam** [comment]
> So without allowing the user to decide which options should be prettier defaults, we can't accurately decide which ones editorconfig should assign over.

`prettier-atom`'s settings are the Prettier defaults by default, so we don't need to force them to set anything. It's also explicit in the Prettier documentation that the defaults are the final fallback and are opinionated.

> Even if we did allow the user to select defaults, prettier doesn't expose its method of merging prettier options with editorconfig options meaning that code would have to be maintained both in this plugin as well as prettier.

This is exactly what [`prettier#resolveConfig`](https://prettier.io/docs/en/api.html#prettierresolveconfigfilepath-options) does when you provide the `editorconfig` option.

**Comment by SavePointSam** [comment]
> It only gets confusing when a prettier config explicitly conflicts with an editorconfig.

Prettier configs are always the highest priority.

File: src/executePrettier/buildPrettierOptions.js (hunk 5 above)
Comment (SavePointSam) [reply src/executePrettier/buildPrettierOptions.js]: I would suggest moving the Prettier formatting settings into their own block to be ignored if a configuration was already detected and then merge in the parser settings all the time as Prettier should be told which language it is working against, which we know from the editor. :D

**Comment by robwise** [comment]
> It seems my thoughts on the logic here may have been lost in the shuffle. We are still only using the prettier-atom settings in specific use-cases right? (I haven't had time to review the PR, just this conversation.)

@SavePointSam On the contrary, I'm under the impression that this PR is implementing what you were suggesting. Previously we were merging prettierrc and atom settings together, but now we ignore atom settings completely if there is a prettierrc.

File: src/executePrettier/buildPrettierOptions.test.js (hunk 6 above)
Comment (robwise) [reply src/executePrettier/buildPrettierOptions.test.js]: @SavePointSam but then we won't know what was from the editorconfig (which we want to keep) and which is just a prettier default (which we want to ignore)?

**Comment by robwise** [comment]
@SavePointSam can you give us an example return of `resolveConfig` when `editorconfig` is true? Do the editorconfig settings come in a separate key than the prettier config settings? Or are they merged together? If the latter, we lose the ability to know whether any given setting was determined from the prettierrc or the editorconfig.

**Comment by SavePointSam** [comment]
> can you give us an example return of resolveConfig when editorconfig is true? 

It will merge them into the final payload. They are not isolated in any way. However, we know that only these are supported: `indent_style` `indent_size/tab_width` `max_line_length`. If we need the ability to know which properties came from where, I can write up a proposal to extend the API.

**Comment by olsonpm** [comment]
> prettier-atom's settings are the Prettier defaults by default, so we don't need to force them to set anything.

That's problematic for two reasons
1. As soon as prettier is updated along with its defaults, ambiguity ensues.
2. If we are to stay inline with Prettier's compilation of `prettier config` > `editor config` > `prettier defaults`, and if we assume settings which match prettier defaults should be overwritten by those found in `editor config`, then a user has no way of editing the atom prettier settings to override it.
<br>

> This is exactly what prettier#resolveConfig does when you provide the editorconfig option.
That's not what I meant.  However looking into it, seems prettier now depends on [editorconfig-to-prettier](https://github.com/josephfrazier/editorconfig-to-prettier) which this plugin could do as well instead of [rolling its own](https://github.com/prettier/prettier-atom/blob/master/src/executePrettier/buildEditorConfigOptions.js).  Still I think the ideal scenario would be for prettier to expose a function `prettier.format(string [, prettierConfig [, editorConfig]])` and for the plugin to rely on that instead of merging the options itself, though that's very minor.
<br>

> Prettier configs are always the highest priority.
Functionally yes that's how it works.  But any editorconfig tool that isn't prettier-aware will not be in line with how prettier formats the code and be awkward as a result (I don't use editorconfig, but I imagine it'd be similar to conflicting linter settings where red lines would show on tabs because it doesn't realize prettier is configured to override them to spaces).
<br>

> If we need the ability to know which properties came from where, I can write up a proposal to extend the API.
I think #178 would find that beneficial.

File: src/executePrettier/buildPrettierOptions.js (hunk 5 above)
Comment (olsonpm) [reply src/executePrettier/buildPrettierOptions.js]: Sounds good.  Sorry for overlooking that.

File: src/executePrettier/buildPrettierOptions.test.js (hunk 6 above)
Comment (olsonpm) [reply src/executePrettier/buildPrettierOptions.test.js]: This is the point I was addressing in my comment on the thread below.  Because there's no way for a user to declare a default in the plugin settings, we can't possibly have a 1-to-1 solution matching prettier's behavior.

**Comment by olsonpm** [comment]
Alright I just pushed an update which fixes the scope-specific settings as well as organizes the tests properly.

For the scope specific settings I left in the [json's `trailingComma`](https://github.com/prettier/prettier-atom/compare/master...olsonpm:use-only-prettier-config-when-found#diff-440e1b1be739422d3e188741e0e71205R53) even though it's not a parser setting.  I assumed this is correct but let me know.

**Comment by SavePointSam** [comment]
> As soon as prettier is updated along with its defaults, ambiguity ensues.

You're not wrong here. We should add a 'use Prettier default' option to each config setting, not assume Prettier defaults in the event they change.

> If we are to stay inline with Prettier's compilation of prettier config > editor config > prettier defaults, and if we assume settings which match prettier defaults should be overwritten by those found in editor config, then a user has no way of editing the atom prettier settings to override it.

This is why `prettier-atom` settings should go after `.editorconfig`, but before actual prettier defaults. Then you don't have this issue as `.editorconfig` has a higher precedent.

> That's not what I meant. However looking into it, seems prettier now depends on editorconfig-to-prettier which this plugin could do as well instead of rolling its own.

Absolutely! Please do this.

> Still I think the ideal scenario would be for prettier to expose a function prettier.format(string [, prettierConfig [, editorConfig]]) and for the plugin to rely on that instead of merging the options itself, though that's very minor.

This puts too much power in the consumers hand, IMO. It will lead to issues that Prettier shouldn't need to support. What happens when an implementation of prettier as an editor plugin has flawed/inconsistent logic of how to merge the options? The Prettier doc say it should do this, but it doesn't. Leaving the configuration resolution in Prettier ensures a cohesive and simple experience for all and prevents various issues that come with more advanced options. It also follows Prettier's stance on being opinionated and normalizes config resolution logic. We have to keep in mind there is a CLI that needs to be supported to, so there will need to be opinionated config resolution regardless of how we consume Prettier.

> But any editorconfig tool that isn't prettier-aware will not be in line with how prettier formats the code and be awkward as a result

This is the end user's responsibility. We can't support conflicting plugins/ideologies.

> I think #178 would find that beneficial.

Sounds good to me! I may take a look once my other WIP stuff is complete. Been rather busy with work and such this past week.

> For the scope specific settings I left in the json's trailingComma even though it's not a parser setting.

Right call! 👍

---

Everything looks good to me. We need to revisit how `buildEditorConfigOptionsIfAppropriate` works, though that falls in line with my current work and is out of the scope of these changes.

Thanks for the work on this!

**Comment by olsonpm** [comment]
> Absolutely! Please do this.

I can tackle this in the current PR if you guys want, otherwise it would make more sense for me to wait for this to be merged and branch from that.
<br>

> This puts too much power in the consumers hand, IMO. It will lead to issues that Prettier shouldn't need to support.

Ah you're right 👍

**Comment by olsonpm** [comment]
One [additional minor change](https://github.com/prettier/prettier-atom/pull/374/commits/46d9f21ac3e0f2e4512f009bdf54ed54b77f5a26#diff-ef0cf7f7d1de7ad6f00d1190065badefL171)

Just removed the incorrect note for the `isDisabledIfNoConfigFile` option:

> *Note:** You must have at least one valid rule in order for your config to be considered present.

**Comment by SavePointSam** [comment]
Let's refactor the .editorconfig parser/reconciler in a seperate PR. :D small scopes!

**Comment by robwise** [comment]
@olsonpm This PR still has WIP in the title, is that still true or okay for me to merge?

**Comment by robwise** [comment]
Also, can we consider this as resolving #218 as well?

**Comment by olsonpm** [comment]
Thanks for checking in with me.  I wanted to thoroughly test it manually before letting it be merged.  I wanted first to make sure an implementation was landed on before I spent a good amount of time manually testing.

That thread is a big one haha, give me a minute.

**Comment by olsonpm** [comment]
Alright it does seem that issue should be resolved with this commit.  I'll confirm though via manual testing.

**Comment by olsonpm** [comment]
Ok - at first I was expecting eslint to follow a similar order to editorconfig e.g. `prettier config` > `eslint config` > `prettier defaults`.  Seems it has its own way of figuring out things especially with the `prettier-atom` setting "Run Prettier Last".  I'm thus assuming the order is fine considering we didn't modify it.

I finished manually testing the following combinations:
- no configs
- .prettierrc
- .editorconfig
- .eslintrc
- .prettierrc and .editorconfig
- .prettierrc and .eslintrc

I didn't test scopes, just a couple settings to prove they were being grabbed/ignored/overridden according to toggles

Anyhoo, this feels ready for me.  If anything goes wrong I'll make sure to work toward fixes as I have plenty of time.

**Comment by robwise** [comment]
@olsonpm SGTM! Awesome job!
